File: 146-lru-cache/146-lru-cache.cs
public class LRUCache {
    public Dictionary<int, Node> map;
    public int cap;
    public int currlen;
    public Node head;
    public Node tail;
    public LRUCache(int capacity) {
        map = new Dictionary<int,Node>();
        cap = capacity;
        this.head = new Node();
        this.tail = new Node();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
    
    public int Get(int key) {
        if(map.ContainsKey(key)){
            Node curr = map[key];
            
            //removing the node from linked list
            curr.prev.next = curr.next;
            curr.next.prev = curr.prev;
            
            //adding on front of linked list
            curr.next = this.head.next;
            this.head.next.prev = curr;
            this.head.next = curr;
            curr.prev = this.head;
            
            //update in map again
            this.map[key] = curr;
            
            return curr.val;
        }
        
        return -1;
    }
    
    public void Put(int key, int value) {
        if(!map.ContainsKey(key)){
            this.currlen++; //increasing current length
            
            //adding on front of linked list
            Node curr = new Node(key, value);
            curr.next = this.head.next;
            this.head.next.prev = curr;
            this.head.next = curr;
            curr.prev = this.head;
            
            this.map.Add(key, curr);
            
            if(this.currlen > this.cap){
                //remove from dictionary
                int evictKey = this.tail.prev.key; 
                this.map.Remove(evictKey);
                
                //evict - remove the node just before tail
                this.tail.prev = this.tail.prev.prev;
                this.tail.prev.next = this.tail;
                
                this.currlen--;
            }
        }
        else{
            //Call get just to remove the node from its original position and put it in front of linked list
            Get(key);
            //now just need to update its value
            this.head.next.val = value;
            //update dictionary
            this.map[key] = this.head.next;
        }
    }
}

public class Node{
    public Node next;
    public Node prev;
    public int key;
    public int val;
    public Node(int _key = 0, int _val = 0, Node _next = null, Node _prev = null){
        this.key = _key;
        this.val = _val;
        this.next = _next;
        this.prev = _prev;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */

File: 146-lru-cache/146-lru-cache.java
class LRUCache {
    public ConcurrentHashMap<Integer,Node> map;
    public int cap;
    public int size;
    public DList list;
    public LRUCache(int capacity) {
        map = new ConcurrentHashMap<>();
        cap = capacity;
        size = 0;
        list = new DList();
    }
    
    public int get(int key) {
        if(map.containsKey(key)){
            Node node = map.get(key);
            
            list.Remove(node);
            list.Add(node);
            
            map.put(key, node);
            
            return node.val;
        }
        else
            return -1;
    }
    
    public void put(int key, int value) {
        if(map.containsKey(key)){
           Node node = map.get(key);
            node.val = value;
            
            list.Remove(node);
            list.Add(node);
            
            map.put(key, node);
        }
        else{
            if(size == cap){
                Node removed = list.RemoveLast();
                size--;
                map.remove(removed.key);
            }
            
            Node node = new Node();
            node.key = key;
            node.val = value;
            
            list.Add(node);
            
            map.put(key, node);
            
            size++;
        }
    }
}

 class DList{
    public Node head;
    public Node tail;
    
    public DList(){
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    
    public void Add(Node node){
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }
    
    public void Remove(Node node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
    
    public Node RemoveLast(){
        Node node = tail.prev;
        Remove(tail.prev);
        return node;
    }
    
}

 class Node{
    public int key;
    public int val;
    public Node next;
    public Node prev;
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */